Adding Fixes and functionalities

tiagolobao opened the pull request:
Hello! Here's my contribution. Tell me if you think any other fix is needed or any patter was mismatched :)

Fixes: 
* Game crashes after non-existing map input 102d09c
* Removing mod interface when at title menu 95a025f

Functionalities
* Autocomplete Hints 2d6ca1f
* Teleport pressing enter d8dd366
* Added interface toggle to hide the inputs and make it cleaner 35d130a

Other commits where in progress changes & some other details

tiagolobao committed:
[Bug Fix] invalid map input

tiagolobao committed:
Improve iteraction for map input

tiagolobao committed:
remove teleporter from Title Screen

tiagolobao committed:
Add teleportPosition option handler

tiagolobao committed:
Code formating

tiagolobao committed:
Add Interface toggle

tiagolobao committed:
Add autocomplete

tiagolobao committed:
fix returning to title

tiagolobao committed:
Game states handling improved

tiagolobao committed:
Update README

dmitmel commented on prestart.js:
```diff
@@ -1,43 +1,359 @@
+		 * @param {string} d start directory
+		 */
+		function getJsonFileNamesRecursive(d){
+			fs.readdir(d, (err, files) => {
```
`err` is not handled here

dmitmel commented on prestart.js:
```diff
@@ -1,43 +1,359 @@
+		function getJsonFileNamesRecursive(d){
+			fs.readdir(d, (err, files) => {
+				files.forEach(file => {
+					if( file.search('.json') > 0 ){
```
This probably should've been `file.endsWith('.json')`

dmitmel commented on prestart.js:
```diff
@@ -1,43 +1,359 @@
+						mapValues.push( map.replace(dir,'').substr(1) );
+					}
+					else{
+						getJsonFileNamesRecursive(d+'/'+file);
```
Please, use `path.join` instead of appending slashes

dmitmel commented on prestart.js:
```diff
@@ -1,43 +1,359 @@
+		 * @param {string} marker Map position to start 
+		 */
+		function teleportIfExists(map,marker){
+			mapPath = map.toPath(ig.root + 'data/maps/', '.json');	//From 
```
Note that `path.join` is not required here because `mapPath` is a URL, not an FS path.

dmitmel commented on prestart.js:
```diff
@@ -1,43 +1,359 @@
+				url: mapPath,
+				success: (a) => {
+					console.log(`Teleported to ${map}`)
+					cc.ig.gameMain.teleport(map,setTeleportPosition(marker));
```
Don't use deprecated CCLoader APIs. `cc.ig.gameMain.teleport` is result of a few deprecated features, you should use `ig.game.teleport` in this case.

dmitmel commented on prestart.js:
```diff
@@ -1,43 +1,359 @@
+		 */
+		function setTeleportPosition(marker){
+			return ig.TeleportPosition.createFromJson({
+				marker: markerValues.find( m => m == marker),
```
`markerValues.includes(marker) ? marker : null` communicates the purpose better

dmitmel commented on prestart.js:
```diff
@@ -1,43 +1,359 @@
+			inp.addEventListener("keydown", function (e) {
+				let x = document.getElementById(this.id + "autocomplete-list");
+				if (x) x = x.getElementsByTagName("div");
+				if (e.keyCode == 40) { // DOWN
```
`ig.KEY.DOWN_ARROW` is 40

dmitmel commented on prestart.js:
```diff
@@ -1,43 +1,359 @@
+				if (e.keyCode == 40) { // DOWN
+					currentFocus++;
+					addActive(x);
+				} else if (e.keyCode == 38) { //UP
```
`ig.KEY.UP_ARROW` is 38

dmitmel commented on prestart.js:
```diff
@@ -1,43 +1,359 @@
+				} else if (e.keyCode == 38) { //UP
+					currentFocus--;
+					addActive(x);
+				} else if (e.keyCode == 13) { // ENTER
```
`ig.KEY.ENTER` is 13

dmitmel commented on prestart.js:
```diff
@@ -1,43 +1,359 @@
+				sc.Model.addObserver(sc.model, this);
+			},
+			modelChanged(model, event) {
+				if( model instanceof sc.GameModel ){
+					if(event == sc.GAME_MODEL_MSG.STATE_CHANGED || sc.GAME_MODEL_MSG.SUB_STATE_CHANGED ){
```
These two `if`s can be merged:

```js
if( model instanceof sc.GameModel && (event == sc.GAME_MODEL_MSG.STATE_CHANGED || sc.GAME_MODEL_MSG.SUB_STATE_CHANGED) ){
```

dmitmel commented on prestart.js:
```diff
@@ -1,43 +1,359 @@
+				document.body.insertAdjacentHTML('beforeend',`
+					<img
+						id="interfaceToggle"
+						src="${'teleportIcon'.toPath(ig.root + 'mods/cc-teleport/', '.png')}" 
```
You should probably create the `img` tag by hand (`new Image()` which returns `HTMLImageElement`, then set properties of the said image element etc) instead of passing a literal string into HTML, this opens path to XSS injections.

dmitmel reviewed (changes requested):
Just a few code style comments here and there. Regarding the stock icon you used, here's an optional challenge for you: open it in some image editor (GIMP, Krita etc), add a layer and make it semi transparent, select and configure (i.e. set the stroke width and color) the "Brush" tool and redraw the icon over the stock image. This both avoids copyright issues and won't look like a creative solution.

Sandorik00 commented:
Just a suggestion, you can a bit improve this for icon cuz i'm started but too lazy to finish it.
![CCTeleport_v0](https://user-images.githubusercontent.com/34166906/81488920-2ef8a780-9278-11ea-82e1-87b97a3be803.png)

tiagolobao committed:
err is not handled at getJsonFileNamesRecursive

from PR discussion:
https://github.com/CCDirectLink/cc-teleport/pull/3#discussion_r422564827

tiagolobao committed:
FileString manipulation with endsWith and path.join- Please, use path.join instead of appending slasheshttps://github.com/CCDirectLink/cc-teleport/pull/3#discussion_r422564904- This probably should've been file.endsWith('.json')https://github.com/CCDirectLink/cc-teleport/pull/3#discussion_r422564867

tiagolobao committed:
Remove deprecated code usage

"Don't use deprecated CCLoader APIs. cc.ig.gameMain.teleport is result of a few deprecated features,
you should use ig.game.teleport in this case."
https://github.com/CCDirectLink/cc-teleport/pull/3#discussion_r422565021

tiagolobao committed:
Using ig constants at autocomplete function

- https://github.com/CCDirectLink/cc-teleport/pull/3#discussion_r422565248
- https://github.com/CCDirectLink/cc-teleport/pull/3#discussion_r422565255
- https://github.com/CCDirectLink/cc-teleport/pull/3#discussion_r422565274

tiagolobao committed:
merging ifs

https://github.com/CCDirectLink/cc-teleport/pull/3#discussion_r422565397

tiagolobao committed:
Avoiding XSS injections from image tag

- https://github.com/CCDirectLink/cc-teleport/pull/3#discussion_r422565520

tiagolobao committed:
Pixel art icon

- https://github.com/CCDirectLink/cc-teleport/pull/3#issuecomment-626259476

tiagolobao committed:
Using fs promises

tiagolobao commented:
@dmitmel I made all the changes (Including changing to fs promises).
Sorry for the bad icon... I really suck at pixel art '-'

dmitmel commented:
The code looks good to me, how does this look in the game? You might as well update the screenshot in the README while you are at it (I recommend taking them at 2x pixel size in windowed mode and then cutting everything except CC including window borders).

tiagolobao commented:
I'll use those two screenshots then...

![cc-teleport-mod-view](https://user-images.githubusercontent.com/15076325/81508302-f1455e80-930b-11ea-8341-47827f18b40d.png)
![cc-teleport-mod-view-2](https://user-images.githubusercontent.com/15076325/81508457-ea6b1b80-930c-11ea-815d-eb258268074e.png)

I can also put some background like this:

![cc-teleport-mod-view-3](https://user-images.githubusercontent.com/15076325/81508519-37e78880-930d-11ea-9efc-2acdcec403a7.png)

dmitmel commented:
Could you please add them to README?

tiagolobao committed:
Update Screenshots

tiagolobao commented:
Done!

dmitmel reviewed (approved):
Well, everything LGTM, cc @2767mr
